src/openrct2-ui/windows/Staff.cpp

```diff
@@ -119,1240 +121,1169 @@ static rct_widget* window_staff_page_widgets[] = {
+                OverviewToolAbort(widgetIndex);
+                break;
+        }
     }
 
-    if (w->height < w->min_height)
+    void OnViewportRotate() override
     {
-        w->height = w->min_height;
-        w->Invalidate();
+        switch (page)
+        {
+            case WINDOW_STAFF_OVERVIEW:
+                OverviewViewportRotate();
+                break;
+        }
     }
 
-    if (w->height > w->max_height)
+    void OnTextInput(rct_widgetindex widgetIndex, std::string_view text) override
     {
-        w->Invalidate();
-        w->height = w->max_height;
+        switch (page)
+        {
+            case WINDOW_STAFF_OVERVIEW:
+                OverviewTextInput(widgetIndex, text);
+                break;
+        }
     }
 
-    rct_viewport* viewport = w->viewport;
-    if (viewport != nullptr)
+private:
+#pragma region Common events
+    /**
+     *
+     *  rct2: 0x006BDF55, 0x006BE7DB, 0x006BEBCF
+     */
+    void CommonMouseUp(rct_widgetindex widgetIndex)
     {
-        int32_t new_width = w->width - 30;
-        int32_t new_height = w->height - 62;
-
-        // Update the viewport size
-        if (viewport->width != new_width || viewport->height != new_height)
+        switch (widgetIndex)
         {
-            viewport->width = new_width;
-            viewport->height = new_height;
-            viewport->view_width = viewport->zoom.ApplyTo(new_width);
-            viewport->view_height = viewport->zoom.ApplyTo(new_height);
+            case WIDX_CLOSE:
+                Close();
+                break;
+            case WIDX_TAB_1:
+            case WIDX_TAB_2:
+            case WIDX_TAB_3:
+                SetPage(widgetIndex - WIDX_TAB_1);
+                break;
         }
     }
 
-    WindowStaffViewportInit(w);
-}
-
-/**
- * Handle the dropdown of patrol button.
- *  rct2: 0x006BDF98
- */
-void WindowStaffOverviewMousedown(rct_window* w, rct_widgetindex widgetIndex, rct_widget* widget)
-{
-    switch (widgetIndex)
+    /**
+     *
+     *  rct2: 0x006BE9E9
+     */
+    void CommonPrepareDrawBefore()
     {
-        case WIDX_LOCATE:
-            WindowStaffShowLocateDropdown(w, widget);
-            break;
-        case WIDX_PATROL:
+        ColourSchemeUpdateByClass(this, static_cast<rct_windowclass>(WC_STAFF));
+
+        if (window_staff_page_widgets[page] != widgets)
         {
-            // Dropdown names
-            gDropdownItems[0].Format = STR_SET_PATROL_AREA;
-            gDropdownItems[1].Format = STR_CLEAR_PATROL_AREA;
+            widgets = window_staff_page_widgets[page];
+            InitScrollWidgets();
+        }
+        SetPressedTab();
 
-            auto dropdownPos = ScreenCoordsXY{ widget->left + w->windowPos.x, widget->top + w->windowPos.y };
-            int32_t extray = widget->height() + 1;
-            WindowDropdownShowText(dropdownPos, extray, w->colours[1], 0, 2);
-            gDropdownDefaultIndex = 0;
+        disabled_widgets = 0;
```

Comment:
Remove, instead use `SetDisabledWidget`

Reply:
Isn't it easier to leave it as is instead of having to loop through all widgets to enable them?

Reply:
The problem is that it restricts the number of widgets to 64. The methods allow us to change the underlying implementation without needing change the windows.